Add Prototype for Checking User Login Status

lolfuljames opened the pull request:
https://jtan-sps-summer20.el.r.appspot.com/

Effective changes:
- New comment layout style
- Login required to post comment, logout button available once user logins
- Comments now include email of original poster

lolfuljames committed:
Add Prototype for Checking User Login Status

This commit adds a log message that shows the current user login status
in the console whenever the comment button is pressed.

lolfuljames committed:
Add Prototype for Checking User Login Status

This commit adds a log message that shows the current user login status
in the console whenever the comment button is pressed.

lolfuljames committed:
Merge branch 'week3' of https://github.com/lolfuljames/portfolio-software-product-sprint into week3

lolfuljames committed:
Refactor AccountServlet

This commit removes unneeded comments and imports.

henryken commented on portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:
```diff
@@ -0,0 +1,43 @@
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login-status")
+public class AccountServlet extends HttpServlet {
```
Would suggest to name this servlet according to the endpoint, e.g. LoginStatusServlet

lolfuljames replied:
Done, I was having trouble naming this as well haha

henryken commented on portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:
```diff
@@ -0,0 +1,43 @@
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userLoggedIn;
+    response.setContentType("text/html");
```
It's more common to have this at the bottom of the method and have the actual business logic at the top.

lolfuljames replied:
I don't quite understand this, do you mean the declaration of the variables?

henryken replied:
Sorry I wasn't clear.
What I mean is something like below.
```java
<your business logic>

response.setContentType("text/html");
PrintWriter out = response.getWriter();
out.println(userLoggedIn);
```

lolfuljames replied:
Ah alright

henryken commented on portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:
```diff
@@ -0,0 +1,43 @@
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (!userService.isUserLoggedIn()) {
+      userLoggedIn = "notLogged";
```
This is one way you can do to return from the server, which is using text.
I would like to challenge you to make this more REST API compliant.
If user is logged in, return HTTP 200, else return HTTP 401.

henryken commented on portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:
```diff
@@ -0,0 +1,43 @@
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userLoggedIn;
+    response.setContentType("text/html");
```
It's not right to set content type as HTML, while you're not returning HTML compliant payload.

lolfuljames replied:
Right, I will change it to use the HTTP status instead

henryken commented on portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:
```diff
@@ -0,0 +1,43 @@
+    PrintWriter out = response.getWriter();
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (!userService.isUserLoggedIn()) {
```
How do you make a user logs in to your portfolio website?

lolfuljames replied:
This was part of the first pull request as shown in the guide, I will be sending a new pull request once I've reach an appropriate milestone instead so that the testing is a bit more realistic and feasible.

lolfuljames replied:
In my commit later on, a login button will appear instead of the form.

henryken commented on portfolio/src/main/webapp/script.js:
```diff
@@ -72,3 +75,13 @@ function showComments(data) {
 }
+
+function processLoginStatus(status) {
+  if (status === "Logged") {
```
Let's make this more realistic use case.
If the user is logged in, then he/she can submit a new comment.
Else he/she can only see comments.

lolfuljames committed:
Add Authentication Feature for Comments Section

This commit:
- Adds Authentication Features using Google App Engine UserService.
- Modifies comment input checking
- Adds login/logout button to comments tab
- Modifies the style of comments

Effective changes:
- New comment layout style
- Login required to post comment, logout button available once user logins
- Comments now include email of original poster

lolfuljames committed:
[merge] User Authentication for Posting Comments

This commit:
- Adds Authentication Features using Google App Engine UserService.
- Modifies comment input checking
- Adds login/logout button to comments tab
- Modifies the style of comments

Effective changes:
- New comment layout style
- Login required to post comment, logout button available once user logins
- Comments now include email of original poster

henryken commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,13 +61,16 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     long timestamp = System.currentTimeMillis();
-    if (userComment.isEmpty()) return;
+    String userEmail = userService.getCurrentUser().getEmail();
+    if (userComment.replaceAll("\\s","").isEmpty()) return;
```
Won't this replace any whitespace including in between words?
Maybe you want to use trim() instead?

lolfuljames replied:
Yes, however since this is just checking whether the submitted comment is empty, it should not make a difference. However, i can make it so that it trims first before posting as well, since I do not do any processing on the comment

henryken replied:
Ah I think I was confused due to the replaceAll.
I think the best alternative is to use .trim().isEmpty()?

henryken commented on portfolio/src/main/java/com/google/sps/servlets/LoginStatusServlet.java:
```diff
@@ -0,0 +1,47 @@
+    Gson gson = new Gson();
+
+    if (!userService.isUserLoggedIn()) {
+      userLoggedIn = gson.toJson("notLogged");
```
You actually don't need to send anything in the body anymore.
I think you can just use the response status code to deduce.

lolfuljames replied:
Yeah, I was actually wondering about this. However I read an article from https://www.javamex.com/tutorials/servlets/http_status_code.shtml that recommends we send something in the body to prevent some server info from being sent unintentionally. What do you think?

henryken replied:
I'm not familiar with that rule. Maybe you can give a try and observe whether a message body is automatically inserted?

lolfuljames replied:
Removed it in commit f3148e2, I checked it out, I didn't notice any difference.

henryken replied:
Thanks for checking it out!

henryken commented on portfolio/src/main/webapp/script.js:
```diff
@@ -72,3 +74,39 @@ function showComments(data) {
+  else if (status == 401) {
+    hideCommentForm();
+  }
+  
```
Don't introduce a newline here since it may not be obvious to the reader

lolfuljames committed:
Refactor Comments Tab

This commit changes the following:
- Removes unneeded response text tag when checking login status
- Use trim instead of replaceAll to trim off side whitespace

henryken reviewed (approved):
LGTM
